### ISSUE TITLE
Ability to create a cache for a different job

### ISSUE THREAD

**Issue by fidsamurai-crelio** [jenkinsci/jobcacher-plugin#226]
### What feature do you want to see added?

We have a use case where we build 15 modules on react, the steps and codebase are the same, only the name of the module changes, we run all 15 in parallel using the build jobs declarative step.

What I'd like to request is the ability to run a "cache generator" pipeline which will use the steps like yarn install to create the cache and then save it for use by another job.

**Comment by jonesbusy** [comment]
I have same requirement here. For cache storage optimization and don't multiply 

One job is creating cache (it's a main Jenkins instance). Other instances can consume (read-only) cache.

Ideally we should be able to override the 

- Prefix added in https://github.com/jenkinsci/jobcacher-plugin/pull/263
- The job name

**Comment by repolevedavaj** [comment]
I am thinking if it would make sense to introduce a new step or even a new plugin which is able to handle that in a better way. More similar to the Github cache action but more flexibility. The current api gets more and more complicated and having a completely different step in the same plugin is strange imho.. what do you think @jonesbusy ?

**Comment by jonesbusy** [comment]
Is right that the cache steps is complicated.

It could be a new plugin or new step I don't have strong opinion about that. 

The annoying point is that you will ask users to migrate but at least start from a clean codebase that address most issues we have with jobcacher

**Comment by repolevedavaj** [comment]
I will check how we could achieve that. Maybe there is an elegant way :)

**Comment by jonesbusy** [comment]
Perhaps, I think we are not far to use jobcacher at large scale. My pain point is now that I need to optimize the storage. 

I did it already per job using skipSave and skipRestore parameter depending on the branch, but now I would like to do it across job and even instances (by setting a read-only ACL on specific folders of the S3 storage)

**Comment by repolevedavaj** [comment]
A solution to the initial problem here is to allow storing a cache instance-wide and then add this location as an additional candidate when restoring. I assume this will not solve your issue @jonesbusy as you want to use a cache across multiple instances. I don't know whether it makes sense to handle that in the plugin. Maybe we could allow specifying a global readonly cache item storage?

**Comment by jonesbusy** [comment]
Yes that's correct. Overidding the prefix and job name is not enough. We need to restore or save on different location.

So perhaps an optional global configuration for a general read only cache could work.

**Comment by klerdos** [comment]
Hi, any updates on implementing this request?

In our setup, we have a job that builds feature branches in Git as part of Merge Request validation. Once the MR is merged, a different job runs on the main branch. Since we use the fast-forward merge strategy, the main branch job runs on the same commit as the feature branch job, so we'd like to reuse the cache from the previous run.

Would it be possible to customize the cache location in S3 so that it's not strictly tied to the job name?

**Comment by repolevedavaj** [comment]
Theoretically, we could add a flag to the cache storage to make it non-job specific. What we have to keep in mind is that this could lead to a concurrency issue if multiple jobs are writing to the same location at the same time. We could solve that by introducing a locking mechanism (something like https://plugins.jenkins.io/lockable-resources/).

Probably it does not make sense to combine such a setup with the `cacheValidityDecidingFile`, as the jobs would "fight" against each other.

Another thing is maybe that not all jobs should use non-job specific cache storages. That could be solved by allowing to configure multiple cache storages which can then be configured in the cache step.

**Comment by klerdos** [comment]
Yes, we would use this new flag only for certain jobs that meet specific prerequisites. In our environment, we have several types of Jenkins pipeline jobs. Some run completely independently, while others are grouped into folders and build from a single repository.

As for potential cache conflicts where multiple jobs might write to the same location, I assumed this was already handled. Currently, when we run a single job (which can trigger multiple builds in parallel) and each build uses the same cache, everything works smoothly without any issues.

**Comment by meiswjn** [comment]
Not sure if this is the case already now, but wouldn't it have security implications on shared controllers if one job could change another job's cache?
Perhaps have a flag which disallows changes from other jobs / allow specifying the scope.
